packages/x-data-grid-pro/src/hooks/features/serverSideData/utils.ts

```diff
@@ -0,0 +1,120 @@
+import { GridRowId } from '@mui/x-data-grid';
+import { GridPrivateApiPro } from '../../../models/gridApiPro';
+
+// Make these configurable using dedicated props?
+const MAX_CONCURRENT_REQUESTS = Infinity;
+const QUEUE_PROCESS_INTERVAL_MS = 300;
+
+export const runIfServerMode = (modeProp: 'server' | 'client', fn: Function) => () => {
+  if (modeProp === 'server') {
+    fn();
+  }
+};
+
+export enum RequestStatus {
+  INQUEUE,
+  PENDING,
+  SETTLED,
+  UNKNOWN,
+}
+
+/**
+ * Fetches row children from the server with option to limit the number of concurrent requests
+ * Determines the status of a request based on the enum `RequestStatus`
+ * Uses `GridRowId` to uniquely identify a request
+ */
+export class NestedDataManager {
+  private pendingRequests: Set<GridRowId> = new Set();
+
+  private queuedRequests: Set<GridRowId> = new Set();
+
+  private settledRequests: Set<GridRowId> = new Set();
+
+  private api: GridPrivateApiPro;
+
+  private maxConcurrentRequests: number;
+
+  private queueProcessInterval: number;
+
+  private timer?: string | number | NodeJS.Timeout;
+
+  constructor(
+    privateApiRef: React.MutableRefObject<GridPrivateApiPro>,
+    maxConcurrentRequests = MAX_CONCURRENT_REQUESTS,
+    queueProcessInterval = QUEUE_PROCESS_INTERVAL_MS,
+  ) {
+    this.api = privateApiRef.current;
+    this.maxConcurrentRequests = maxConcurrentRequests;
+    this.queueProcessInterval = queueProcessInterval;
+  }
+
+  private processQueue = async () => {
+    if (this.queuedRequests.size === 0) {
+      clearInterval(this.timer);
+      return;
+    }
+    if (this.pendingRequests.size >= this.maxConcurrentRequests) {
+      return;
+    }
+    const fetchQueue = Array.from(this.queuedRequests);
+    for (let i = 0; i < this.maxConcurrentRequests; i += 1) {
+      const nextId = fetchQueue[i];
+      if (!nextId) {
+        clearInterval(this.timer);
+        return;
+      }
+      this.queuedRequests.delete(nextId);
+      this.api.fetchRowChildren(nextId);
+      this.pendingRequests.add(nextId);
+    }
+  };
+
+  public enqueue = async (ids: GridRowId[]) => {
+    ids.forEach((id) => {
+      if (this.pendingRequests.size < this.maxConcurrentRequests) {
+        this.pendingRequests.add(id);
+        this.api.fetchRowChildren(id);
+      } else {
+        this.queuedRequests.add(id);
+      }
+
+      if (this.queuedRequests.size > 0) {
+        if (this.timer) {
+          clearInterval(this.timer);
+        }
+        this.timer = setInterval(this.processQueue, this.queueProcessInterval);
```

Comment:
Why are we running it in intervals?

Reply:
Instead of binding it with some external API event, the idea is to keep polling until all the requests are moved from INQUEUE to PENDING state.

Reply:
Did you consider running `processQueue` after each request status change instead?

If the requests pool is busy, we can't run another request until one of the running ones finishes, right?